### PR TITLE
Expose aggregate file sizes in FileStats

### PR DESCRIPTION
**Pull request by amcasey**
We're not sure that file counts are a good proxy for project size and this will give us more direct insight.

### CODE REVIEW

**Comment by amcasey** [comment]
FYI @minestarks

**Comment by amcasey** [comment]
@DanielRosenwasser I'd be really happy to have this in 3.2 (after I fix the tests, of course).

File: src/server/scriptInfo.ts [hunk 1]
```diff
@@ -68,37 +77,51 @@ namespace ts.server {
...
             Debug.assert(newText !== undefined);
 
             // Reload always has fresh content
             this.pendingReloadFromDisk = false;
 
+            // We only need both this.text and this.fileSize if this.text
+            // is artificially empty because it was too large.
+            // We assume that an empty string passed to reload is not
+            // for a file that was too large to store.
+            this.fileSize = undefined;
```
Comment (sheetalkamat) [inline src/server/scriptInfo.ts]: I don't think you should set this to undefined here. This should happen at reload callsites because that's where the information of actual file size lies.

File: src/server/scriptInfo.ts [hunk 2]
```diff
@@ -68,37 +77,51 @@ namespace ts.server {
...
+
             // If text changed set the text
             // This also ensures that if we had switched to version cache,
             // we are switching back to text.
             // The change to version cache will happen when needed
             // Thus avoiding the computation if there are no changes
             if (this.text !== newText) {
-                this.useText(newText);
+                this.text = newText;
```
Comment (sheetalkamat) [inline src/server/scriptInfo.ts]: Why this change? Can you please keep useText?

File: src/testRunner/unittests/telemetry.ts [hunk 3]
```diff
@@ -215,11 +215,12 @@ namespace ts.projectSystem {
...
             const tsconfig = makeFile("/jsconfig.json", {});
             const et = new TestServerEventManager([tsconfig, file]);
-            et.host.getFileSize = () => server.maxProgramSizeForNonTsFiles + 1;
+            const fileSize = server.maxProgramSizeForNonTsFiles + 1;
+            et.host.getFileSize = () => fileSize;
             et.service.openClientFile(file.path);
             et.getEvent<server.ProjectLanguageServiceStateEvent>(server.ProjectLanguageServiceStateEvent);
             et.assertProjectInfoTelemetryEvent({
-                fileStats: fileStats({ js: 1 }),
+                fileStats: fileStats({ js: 1, jsSize: fileSize }),
```
Comment (sheetalkamat) [inline src/testRunner/unittests/telemetry.ts]: Can you add another test where the text.length is used to get the fileSize in say ts or js?

Comment (amcasey) [reply src/testRunner/unittests/telemetry.ts]: To confirm, you saw the text storage file size tests and you'd like a telemetry file size test as well?

File: src/server/scriptInfo.ts (hunk 1 above)
Comment (amcasey) [reply src/server/scriptInfo.ts]: I don't understand your suggestion.

File: src/server/scriptInfo.ts (hunk 2 above)
Comment (amcasey) [reply src/server/scriptInfo.ts]: I'll restore it but I have to say that I find it very confusing how many ways there are to change the state of this object.

File: src/server/scriptInfo.ts (hunk 1 above)
Comment (amcasey) [reply src/server/scriptInfo.ts]: I moved it into `useText`.

File: src/server/scriptInfo.ts (hunk 2 above)
Comment (amcasey) [reply src/server/scriptInfo.ts]: Added a comment indicating that it is test-only.

File: src/testRunner/unittests/telemetry.ts (hunk 3 above)
Comment (amcasey) [reply src/testRunner/unittests/telemetry.ts]: Added a test.
